tools/atlas/atlas.nim

```diff
@@ -266,6 +275,59 @@ proc gitPull(c: var AtlasContext; p: PackageName) =
   if status != 0:
     error(c, p, "could not 'git pull'")
 
+proc gitTag(c: var AtlasContext; tag: string) =
+  let (_, status) = exec(c, GitTag, [tag])
+  if status != 0:
+    error(c, c.projectDir.PackageName, "could not 'git tag " & tag & "'")
+
+proc incrementTag(lastTag: string; field: SemVerField): string =
+  var startPos =
+    if lastTag[0] in {'0'..'9'}: 0
+    else: 1
+  var endPos = lastTag.find('.', startPos)
+  if ord(field) >= 1:
+    for i in 1 .. ord(field):
+      assert endPos != -1, "last tag '" & lastTag & "' is missing . periods"
+      startPos = endPos + 1
+      endPos = lastTag.find('.', startPos)
+  if endPos == -1:
+    endPos = len(lastTag)
+  let patchNumber = parseInt(lastTag[startPos..<endPos])
+  lastTag[0 ..< startPos] & $(patchNumber + 1) & lastTag[endPos..^1]
+
+proc incrementLastTag(c: var AtlasContext; field: SemVerField): string =
+  let (ltr, status) = exec(c, GitLastTaggedRef, [])
+  if status == 0:
+    let
+      lastTaggedRef = ltr.strip()
+      (lt, _) = osproc.execCmdEx("git describe --tags " & lastTaggedRef)
+      (cc, _) = exec(c, GitCurrentCommit, [])
+      lastTag = lt.strip()
+      currentCommit = cc.strip()
+
+    if lastTaggedRef == currentCommit:
+      warn c, c.projectDir.PackageName, "the current commit '" & currentCommit & "' is already tagged '" & lastTag & '\''
+      lastTag
+    else:
+      incrementTag(lastTag, field)
+  else: "v0.0.1" # assuming no tags have been made yet
+
+proc pushTag(c: var AtlasContext; tag: string) =
+  let (outp, status) = exec(c, GitPush, [tag])
+  if status != 0:
+    error(c, c.projectDir.PackageName, "could not 'git push " & tag & "'")
+  elif outp.strip() == "Everything up-to-date":
+    message(c, "[Info] ", c.projectDir.PackageName, "is up-to-date")
+  else:
+    message(c, "[Info] ", c.projectDir.PackageName, "successfully pushed tag: " & tag)
+
+proc tag(c: var AtlasContext; field: SemVerField) =
+  let oldErrors = c.errors
+  let newTag = incrementLastTag(c, field)
+  if c.errors == oldErrors:
+    gitTag(c, newTag)
+  pushTag(c, newTag)
```

Comment:
If there are errors, should we really do `pushTag` regardless?

Reply:
The "error" is the tag already existing. It will then tell you that either 1. it successfully pushed the tag or 2. the tag has already been pushed

Reply:
So make `incrementLastTag` return an `Option[newTag]` or something. The `c.errors == oldErrors` is kind of a failed experiment. :-)